### PR TITLE
WIP: Passkeys & Minimalism Exercise

### PR DESCRIPTION
**Pull request by devlyn37**
## Description

This next SDK update that includes passkey wallets is a good opportunity simplify and reduce our API surface!

This PR Does Two Things:

### An Exercise in Minimalism

AA has a lot of moving pieces, unless we're ruthless about simplifying & removing things as we go, we'll drown in unnecessary complexity! This PR drops a ton from the SDK, probably too much. The idea here is that the team will review, we'll figure out what's really needed, and drop the rest. I've annotated different pieces in the diff.

Big Questions:
- Does the "session key" setup we have actually work? What's the intention behind it?
- Why are we rolling out own signer methods (`sendTransaction`, `signMessage`, etc) in `EVMAAWallet`? How about clients grab the signer and use that instead?
- Most of our E2E testing here seems to be related to `ethers` or `fireblocks`, both things we don't plan on supporting atm, are they needed anymore?

### Add Passkey Wallet M1 

This PR adds a draft of the API we'll expose for creating passkey wallets. I'll annotate this within the PR!

Things that are totally WIP and not ready for review yet:
- Passkey edge cases & errors we return from the SDK
- The implementation of the SDK
    - I still have this branch using ZD's passkey server
    - We still need to store passkey wallets within the CM backend.

## Test plan

Not sure yet, there will be one though.

### CODE REVIEW

File: packages/client/wallets/aa/src/api/BaseCrossmintService.ts [hunk 1]
```diff
@@ -65,7 +65,7 @@ export abstract class BaseCrossmintService extends LoggerWrapper {
     }
 
     protected getUrlFromEnv(environment: string) {
-        const url = BaseCrossmintService.urlMap[environment];
+        const url = BaseCrossmintService.urlMap["staging"];
```
Comment (devlyn37) [inline packages/client/wallets/aa/src/api/BaseCrossmintService.ts:68]: temp only for testing

File: packages/client/wallets/aa/src/api/CrossmintWalletService.ts
Comment (devlyn37) [inline packages/client/wallets/aa/src/api/CrossmintWalletService.ts:1]: dropping fireblocks ncw related pieces

File: packages/client/wallets/aa/src/CrossmintAASDK.ts [hunk 2]
```diff
@@ -1,151 +1,96 @@
...
-        });
-    }
-
-    private async getEntryPointVersion<B extends EVMBlockchainIncludingTestnet = EVMBlockchainIncludingTestnet>(
-        userIdentifier: UserIdentifierParams,
-        chain: B | EVMBlockchainIncludingTestnet
-    ): Promise<EntryPointVersion> {
-        if (userIdentifier.email == null && userIdentifier.userId == null) {
-            throw new WalletSdkError(`Email or userId is required to get the entry point version`);
+    public async getOrCreateWallet(
```
Comment (devlyn37) [inline packages/client/wallets/aa/src/CrossmintAASDK.ts:50]: Here's the main proposed change for the SDK. Described above in comment

File: packages/client/wallets/aa/src/blockchain/wallets/passkey.ts
Comment (devlyn37) [inline packages/client/wallets/aa/src/blockchain/wallets/passkey.ts:1]: Here's where you can see passkey wallet behavior from the SDK

Comment (devlyn37) [reply packages/client/wallets/aa/src/blockchain/wallets/passkey.ts:1]: Still WIP, but primary ideas are there, pretty simple

File: packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts
Comment (devlyn37) [inline packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts:1]: I'm probably missing observability improvements from the merge, will make sure everything is there again before this is ready to be reviewed for release

File: packages/client/wallets/aa/src/blockchain/wallets/passkey.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import { createPasskeyValidator, deserializePasskeyValidator } from "@zerodev/passkey-validator";
+import { KernelValidator, createKernelAccount } from "@zerodev/sdk";
+import { ENTRYPOINT_ADDRESS_V07 } from "permissionless";
+import { ENTRYPOINT_ADDRESS_V07_TYPE } from "permissionless/_types/types";
+import { HttpTransport, PublicClient } from "viem";
+
+import { EVMBlockchainIncludingTestnet, UserIdentifier } from "@crossmint/common-sdk-base";
+
+export default class PasskeyWalletService {
+    private readonly entryPoint = ENTRYPOINT_ADDRESS_V07;
```
Comment (maxisch) [inline packages/client/wallets/aa/src/blockchain/wallets/passkey.ts:12]: We probably need to add the entry point verification. Apart from backwards compatibility. some chains like astar don't support v07 and only work for v06

File: packages/client/wallets/aa/src/blockchain/wallets/passkey.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        if (serializedData != null) {
+            validator = await deserializePasskeyValidator(publicClient, {
+                serializedData,
+                entryPoint: this.entryPoint,
+            });
+        } else {
+            validator = await createPasskeyValidator(publicClient, {
+                passkeyServerUrl: this.passkeyServerUrl,
+                entryPoint: this.entryPoint,
+                passkeyName: getIdString(userIdentifier),
```
Comment (maxisch) [inline packages/client/wallets/aa/src/blockchain/wallets/passkey.ts:33]: we need to add the omit flag to the credentials header: https://github.com/Crossmint/crossmint-sdk/pull/485/files#diff-44d9925b3eab949d3dc6cce03789d9160cfa1d1a34305326e05e39be67f70920R77

File: packages/client/wallets/aa/playwright/tests/fireblocks/fireblock.spec.ts
Comment (devlyn37) [inline packages/client/wallets/aa/playwright/tests/fireblocks/fireblock.spec.ts:1]: No need to have this if we don't plan on shipping for a while, we can come back to the diff tree later if needed to pick up again.

File: packages/client/wallets/aa/pnpm-lock.yaml
Comment (devlyn37) [inline packages/client/wallets/aa/pnpm-lock.yaml:1]: Ignore this, I'll convert back later

File: packages/client/wallets/aa/src/api/CrossmintWalletService.ts [hunk 5]
```diff
@@ -7,17 +7,6 @@ import { BaseCrossmintService } from "./BaseCrossmintService";
 export { EVMBlockchainIncludingTestnet } from "@crossmint/common-sdk-base";
 
 export class CrossmintWalletService extends BaseCrossmintService {
-    async createSessionKey(address: string | `0x${string}`) {
```
Comment (devlyn37) [inline packages/client/wallets/aa/src/api/CrossmintWalletService.ts:10]: Part of dropping the session key functionality from eoa wallets. Will re-add if someone is confident that it works and/or someone is using.

File: packages/client/wallets/aa/src/blockchain/BlockchainNetworks.ts
Comment (devlyn37) [inline packages/client/wallets/aa/src/blockchain/BlockchainNetworks.ts:1]: temp

File: packages/client/wallets/aa/src/types/Config.ts [hunk 6]
```diff
@@ -58,50 +46,28 @@ export type ViemAccount = {
...
-
-export type DecryptInput = {
-    pkpPublicKey: string;
-    pkpEthAddress: string;
-    cipherText: string;
-    dataToEncryptHash: string;
-    capacityDelegationAuthSig: AuthSig;
-};
-
-type Cipher = { method: "lit_protocol"; data: LitProtocolCipherData };
```
Comment (devlyn37) [inline packages/client/wallets/aa/src/types/Config.ts:93]: Deleting since it's unused for now, can grab from source tree later if needed

File: packages/client/wallets/aa/src/types/Config.ts [hunk 7]
```diff
@@ -58,50 +46,28 @@ export type ViemAccount = {
...
-
-export type BackwardsCompatibleChains = "goerli";
```
Comment (devlyn37) [inline packages/client/wallets/aa/src/types/Config.ts:95]: Do we really need this? The generic it forced us to have everywhere was noisy and made the little TS preview of our classes mostly unusable.

File: packages/client/wallets/aa/src/types/Config.ts [hunk 8]
```diff
@@ -58,50 +46,28 @@ export type ViemAccount = {
...
+export type WalletConfig = EOAWalletConfig | PasskeyWalletConfig;
 
 export type Client = {
     publicClient: PublicClient;
     walletClient: ReturnType<
         typeof createKernelAccountClient<
```
Comment (devlyn37) [inline packages/client/wallets/aa/src/types/Config.ts:66]: order swapped because of update to ZD sdk

File: packages/client/wallets/aa/src/blockchain/BlockchainNetworks.ts [hunk 9]
```diff
@@ -217,8 +217,8 @@ export const getZeroDevProjectIdByBlockchain = logInputOutput((chain: EVMBlockch
         ["optimism", ZD_OPTIMISM_PROJECT_ID],
         ["arbitrum", ZD_ARBITRUM_PROJECT_ID],
         ["ethereum-goerli", ZD_GOERLI_PROJECT_ID],
-        ["ethereum-sepolia", ZD_SEPOLIA_PROJECT_ID],
-        ["polygon-amoy", ZD_AMOY_PROJECT_ID],
+        ["ethereum-sepolia", "1743c9e4-c41a-4188-827f-547301aa1645"],
+        ["polygon-amoy", "96021f51-61b7-4385-b57f-e81f3f4c4f70"],
```
Comment (maxisch) [inline packages/client/wallets/aa/src/blockchain/BlockchainNetworks.ts:221]: reminder to revert

File: packages/client/wallets/aa/src/api/CrossmintWalletService.ts (hunk 5 above)
Comment (maxisch) [reply packages/client/wallets/aa/src/api/CrossmintWalletService.ts:10]: If we decide to drop this, we need to mark it as optional on crossbit-main's endpoint.

File: packages/client/wallets/aa/src/adapters/reservoir.ts [hunk 10]
```diff
@@ -4,19 +4,20 @@ import { hexToBigInt, http } from "viem";
 import { getBundlerRPC } from "../blockchain/BlockchainNetworks";
 import { EVMAAWallet } from "../blockchain/wallets/EVMAAWallet";
 
+// TODO does this still all work?
 export function reservoirAdapter(aaWallet: EVMAAWallet): ReservoirWallet {
     return {
-        address: async () => aaWallet.getAddress() as string,
+        address: async () => aaWallet.getAddress(),
```
Comment (MatiNavCrossmint) [inline packages/client/wallets/aa/src/adapters/reservoir.ts:10]: As tt returns a `0x${string}`, we nedeed to specify the `string` type returned

File: packages/client/wallets/aa/src/adapters/reservoir.ts [hunk 11]
```diff
@@ -4,19 +4,20 @@ import { hexToBigInt, http } from "viem";
 import { getBundlerRPC } from "../blockchain/BlockchainNetworks";
 import { EVMAAWallet } from "../blockchain/wallets/EVMAAWallet";
 
+// TODO does this still all work?
```
Comment (MatiNavCrossmint) [inline packages/client/wallets/aa/src/adapters/reservoir.ts:7]: Yes, you can test it [here](https://github.com/Crossmint/scw_demo/blob/main-w3a/src/app/utils/reservoir.ts)

File: packages/client/wallets/aa/src/utils/blockchain.ts [hunk 12]
```diff
@@ -14,55 +10,6 @@ export type VerifyMessageParams = {
...
- * so we need to make the hexlify fix manually when sending a transaction.
- * That fix can be found here:
- * https://github.com/ethers-io/ethers.js/commit/a12030ad29aa13c02aa75d9e0860f4986a0043b4#diff-047e7ebfdbd5c41e762cda03593bd15ed8b3121dece67262729dfcbde7040818R221
- * And more context on this issue:
- * https://github.com/ethers-io/ethers.js/issues/614
- *
- * Applying to it hexValue is used as a workaround on Crossbit, but doesn't cover all the cases (i.e. 0x71afd498d0000).
- * This function is the same as hexValue but without hexStripZeros, which is what makes it to not work.
- */
-export async function decorateSendTransactionData(transaction: Deferrable<TransactionRequest>) {
```
[the remark this answers is not in the document]
Comment (MatiNavCrossmint) [reply packages/client/wallets/aa/src/utils/blockchain.ts:29]: As far as I know it was an issue, if we decide to remove it then we need to test it as it was a hack to make Txs work

File: packages/client/wallets/aa/src/CrossmintAASDK.ts (hunk 2 above)
Comment (MatiNavCrossmint) [reply packages/client/wallets/aa/src/CrossmintAASDK.ts:48]: Can we re-add the `logPerformance` wrapper?

**Review by MatiNavCrossmint — COMMENTED** [review]
My comments are kind of `request changes` ones, but will just comment to avoid blocking!

**Review by afeight — COMMENTED** [review]
sharing thoughts as I go–I agree w/ Mati that we should think more about ethers support
But I am totally comfortable ripping out Fireblocks

File: packages/client/wallets/aa/pnpm-lock.yaml [hunk 13]
```diff
@@ -1,6435 +0,0 @@
-lockfileVersion: '6.0'
```
Comment (afeight) [inline packages/client/wallets/aa/pnpm-lock.yaml:1]: +1, we've been using `yarn` so far

File: packages/client/wallets/aa/src/adapters/reservoir.ts [hunk 14]
```diff
@@ -28,8 +29,7 @@ export function reservoirAdapter(aaWallet: EVMAAWallet): ReservoirWallet {
         },
         handleSendTransactionStep: async (chainId, stepItem, _) => {
             const stepData = stepItem.data;
-            return await aaWallet.sendTransaction({
-                chainId: chainId,
```
Comment (afeight) [inline packages/client/wallets/aa/src/adapters/reservoir.ts:32]: Not sure, but I think I saw issues where I had to include this–any proof to being able to remove or just saw that it's optional?
Test cases here to confirm would be to use with:
- amoy
- optimism
- astar (if supported by reservoir, I could see this being different)

File: packages/client/wallets/aa/src/passkeys/PasskeysSDK.ts [hunk 15]
```diff
@@ -1,190 +0,0 @@
-import { logError, logInfo, logWarn } from "@/services/logging";
```
Comment (afeight) [inline packages/client/wallets/aa/src/passkeys/PasskeysSDK.ts:1]: yep, unfortunately blocked and related to fireblocks anyways
pour one out 😢🍺

File: packages/client/wallets/aa/src/types/Config.ts
[the remark this answers is not in the document]
Comment (afeight) [reply packages/client/wallets/aa/src/types/Config.ts:1]: yep I'm on board–ideally we never use them anymore

File: packages/client/wallets/aa/src/utils/deferrable.ts [hunk 16]
```diff
@@ -1,14 +0,0 @@
-import { Deferrable } from "ethers/lib/utils";
```
Comment (afeight) [inline packages/client/wallets/aa/src/utils/deferrable.ts:1]: yeah fuck this

File: packages/client/wallets/aa/src/utils/helpers.ts [hunk 17]
```diff
@@ -13,14 +9,3 @@ export function isLocalhost() {
 export function isEmpty(str: string | undefined | null): str is undefined | null {
     return !str || str.length === 0 || str.trim().length === 0;
 }
-
-export function transformBackwardsCompatibleChains<B extends BlockchainIncludingTestnet = BlockchainIncludingTestnet>(
```
[the remark this answers is not in the document]
Comment (afeight) [reply packages/client/wallets/aa/src/utils/helpers.ts:17]: This may have been @maxisch 
I forget, but reading through it only maps goerli which shouldn't be supported anymore anyways. I say remove

File: packages/client/wallets/aa/src/utils/user.ts [hunk 18]
```diff
@@ -39,6 +28,19 @@ export function parseUserIdentifier(reqParams: UserIdentifierParams, acceptPhone
...
+export function getIdString(identifier: UserIdentifier): string {
+    switch (identifier.type) {
+        case "whiteLabel":
+            return identifier.userId;
+        case "email":
+            return identifier.email;
+        case "phoneNumber":
+            return identifier.phoneNumber;
+        default:
+            throw new Error("Invalid UserIdentifier type.");
```
Comment (afeight) [inline packages/client/wallets/aa/src/utils/user.ts:40]: TBH I'd rather omit this and let the compiler fail if we add a case and forget to handle it

File: packages/client/wallets/aa/src/api/CrossmintWalletService.ts (hunk 5 above)
Comment (afeight) [reply packages/client/wallets/aa/src/api/CrossmintWalletService.ts:10]: I cannot confirm it's working, but I wouldn't be comfortable dropping this _yet_
One of the initial reqs for the project was interop with Crossmint.com, at least the form of "I can transfer nfts from crossmint.com"

File: packages/client/wallets/aa/src/api/CrossmintWalletService.ts [hunk 19]
```diff
@@ -41,101 +30,6 @@ export class CrossmintWalletService extends BaseCrossmintService {
         );
     }
 
-    async generateChainData(input: GenerateSignatureDataInput) {
-        return this.fetchCrossmintAPI(
-            "unstable/wallets/aa/wallets/chaindata",
-            { method: "POST", body: JSON.stringify(input) },
-            "Error setting custodian. Please contact support"
-        );
```
Comment (afeight) [inline packages/client/wallets/aa/src/api/CrossmintWalletService.ts:49]: is this also just used in fireblocks? Or is it part of the session key flow?

File: packages/client/wallets/aa/src/api/CrossmintWalletService.ts [hunk 20]
```diff
@@ -41,101 +30,6 @@ export class CrossmintWalletService extends BaseCrossmintService {
...
-    async updateWallet(address: string, enableSig: string, version: number) {
-        return this.fetchCrossmintAPI(
-            "unstable/wallets/aa/wallets/version/update",
-            {
-                method: "POST",
-                body: JSON.stringify({ address, enableSig, version }),
-            },
-            `Error updating wallet: ${address}`
-        );
-    }
```
Comment (afeight) [inline packages/client/wallets/aa/src/api/CrossmintWalletService.ts:137]: let's check product req on these

File: packages/client/wallets/aa/src/api/CrossmintWalletService.ts [hunk 21]
```diff
@@ -41,101 +30,6 @@ export class CrossmintWalletService extends BaseCrossmintService {
...
-
-    async getNCWIdentifier(deviceId: string) {
-        return this.fetchCrossmintAPI(
-            `unstable/wallets/aa/ncw?deviceId=${deviceId}`,
-            { method: "GET" },
-            `Error getting NCW identifier for device: ${deviceId}`
-        );
-    }
```
Comment (afeight) [inline packages/client/wallets/aa/src/api/CrossmintWalletService.ts:115]: just making the note to remove these from server also–they are likely already unused and safe to delete

File: packages/client/wallets/aa/src/CrossmintAASDK.ts [hunk 22]
```diff
@@ -1,151 +1,94 @@
...
+                transport: http(getBundlerRPC(chain)),
+            });
+            const userIdentifier = parseUserIdentifier(user);
+
+            let wallet: EVMAAWallet;
+            if (walletConfig != null && isEOAWalletConfig(walletConfig)) {
+                wallet = await this.eoaWalletService.getOrCreate(userIdentifier, chain, publicClient, walletConfig);
+            } else {
+                wallet = await this.passkeyWalletService.getOrCreate(userIdentifier, chain, publicClient, walletConfig);
+            }
```
Comment (afeight) [inline packages/client/wallets/aa/src/CrossmintAASDK.ts:74]: nice separation 👍

File: packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts [hunk 23]
```diff
@@ -1,406 +1,65 @@
...
-                    chain: evmToken.chain,
-                });
-                throw new TransferError(
-                    `Error transferring token ${evmToken.contractAddress}${tokenId == null ? "" : `:${tokenId}}`}`
-                );
-            }
-        });
-    }
-
-    getSigner<Type extends SignerType>(type: Type): SignerMap[Type] {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts:278]: the assumption we had when startng was that ethers was more likely to be used than viem (at the time true)
It's hard to imagine that won't be a re-emerging req in the future, so I'd make a push to keep the function signature the same–`getSigner("viem")`

File: packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts [hunk 24]
```diff
@@ -1,406 +1,74 @@
...
-                };
-
-                logInfo(`[EVMAAWallet - SEND_TRANSACTION] - tx_params: ${JSON.stringify(txParams)}`);
-
-                return await this.kernelClient.sendTransaction(txParams);
-            } catch (error) {
-                throw new TransactionError(`Error sending transaction: ${error}`);
-            }
-        });
-    }
```
[the remark this answers is not in the document]
Comment (afeight) [reply packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts:196]: yeah tbh this one doesn't seem to add any value

File: packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts [hunk 25]
```diff
@@ -1,406 +1,65 @@
...
-                    },
-                },
-            }),
+            ...(hasEIP1559Support(chain) && { middleware: { sponsorUserOperation } }),
         });
-        this.account = account;
-        this.entryPoint = entryPoint;
-    }
-
-    getPaymasterClient() {
```
Comment (afeight) [inline packages/client/wallets/aa/src/blockchain/wallets/EVMAAWallet.ts:111]: was this used anywhere?
